pmlc/dialect/comp/ir/interfaces.td

```diff
@@ -0,0 +1,80 @@
+// Copyright 2020 Intel Corporation
+#ifndef __PMLC_COMP_INTERFACES__
+#define __PMLC_COMP_INTERFACES__
+
+include "mlir/IR/OpBase.td"
+
+class COMP_OpInterface<string name> : OpInterface<name>;
+
+def COMP_ExecEnvOpInterface : COMP_OpInterface<"ExecEnvOpInterface"> {
+  let description = [{
+    Ties operation to specific execution environment.
+  }];
+
+  let methods = [
+    InterfaceMethod<"", "::mlir::Value", "getExecEnv", (ins), [{}], [{
+      ConcreteOp op = ::mlir::cast<ConcreteOp>(this->getOperation());
+      return op.execEnv();
+    }]>
+  ];
+}
+
+def COMP_ScheduleOpInterface : COMP_OpInterface<"ScheduleOpInterface"> {
+  let description = [{
+    Interface for operations that are scheduled on execution environment.
+    Provides access for dependant event and produced event.
+    Operations implementing this interface are expected to also implement
+    ExecEnvOp interface.
+  }];
+
+  let methods = [
```

Comment:
This seems to be defining a completely orthogonal mechanism for specifying dependencies between operations.  Is there any chance you could use MLIR's SSA arguments and results to do the same thing?  Or does that not work for some reason?  (If so, please add a comment.)

Reply:
I have added short description of advantages of event based approach over SSA dependencies to `EventType` description.